### ISSUE TITLE
where is the backward() function

### ISSUE THREAD

**Issue by hnczwj2008** [pvjosue/pytorch_convNd#1]
Hi josue:
where is the backward() function in your class? Is there an automatically backward() function ,or we need to write it by ourself ?

**Comment by pvjosue** [comment]
Hi,
The backward function is automatically computed by autograd. To use it you  would do like with any other tensor.
First you have to compute a loss and then call backward from the loss like:

```python
import torch
from convNd import convNd

device = "cpu"

# define basic layer info
inChans = 2
outChans = 4
weight = torch.rand(1)[0]
bias = torch.rand(1)[0]

# create input tensor
x = torch.rand(1, inChans, 5, 5, 5, 5, 5).to(device)
conv5d = convNd(
    in_channels=inChans, 
    out_channels=outChans,
    num_dims=5, 
    kernel_size=3, 
    stride=(2,1,1,1,1), 
    padding=0, 
    padding_mode='zeros',
    output_padding=0,
    is_transposed=False,
    use_bias=True, 
    groups=2,
    kernel_initializer=lambda x: torch.nn.init.constant_(x, weight), 
    bias_initializer=lambda x: torch.nn.init.constant_(x, bias)).to(device)
# apply conv5d
xConv = conv5d(x)

# Compute a loss
lossF = xConv.mean()
# Call backward on that loss
lossF.backward()
# check gradients of conv weights
print(conv5d.conv_layers[0].conv_layers[0].weight.grad.sum())
```
As this is a 5D conv, Note how on the final line *conv5d.conv_layers[0]* access the first 4D conv and  *conv5d.conv_layers[0].conv_layers[0]* the first 3D convolution, where the actual Pytorch torch.conv3d is used and the weights are stored.

**Comment by hnczwj2008** [comment]
got it,  thank you !

**Comment by hnczwj2008** [comment]
Hi josue:
I want to realize a 6D-conv-layer. Then I try this:
```py
trainset_dataloader = DataLoader(dataset=train_dataset,
                                 batch_size=8,
                                 shuffle=True,
                                 num_workers=4)
import torch
from pytorch_convNd.convNd import convNd

for epoch in range(1):
    for i, traindata in enumerate(trainset_dataloader):
        trian_image, train_label = traindata
        trian_image_unsque = trian_image.unsqueeze(0)
        print(trian_image_unsque.shape) #==> torch.Size([1, 8, 3, 3, 3, 3, 3, 3])
        device = "cpu"
        # define basic layer info
        inChans = 8 # is this equal to batch_size ?
        outChans = 4
        weight = (torch.rand(1)[0])
        bias = (torch.rand(1)[0])
        # create input tensor
        # x = torch.rand(1, inChans, 5, 5, 5, 5, 5).to(device)
        x = trian_image_unsque.to(device)
        conv6d = convNd(
            in_channels=inChans, 
            out_channels=outChans,
            num_dims=6, 
            kernel_size=3, 
            stride=(1,1,1,1,1,1), 
            padding=0, 
            padding_mode='zeros',
            output_padding=0,
            is_transposed=False,
            use_bias=True, 
            groups=2,
            kernel_initializer=lambda x: torch.nn.init.constant_(x, weight), 
            bias_initializer=lambda x: torch.nn.init.constant_(x, bias)).to(device)
        # apply conv6d
        xConv6d = conv6d(x)
        # Compute a loss
        lossF = xConv6d.mean()
        # Call backward on that loss
        lossF.backward()
        if i==0:
            break
```

First, a single data's shape is 3X3X3X3X3X3,and I regard it as an 6-dimension data. Second, I set batch_size 8, so I got an 8X3X3X3X3X3X3 tensor from DataLoader. Third, I used `trian_image_unsque = trian_image.unsqueeze(0)`  ,then got an 1X8X3X3X3X3X3X3 tensor,which was feed into 6DConv .
Am I right ?  
In your 5D demo, is the "inChans"  has same meaning with "batch_size" ? 
And the first parameter  "1" , which is before "inChans", is an adapting dimension for training ? (code:  x = torch.rand(1, inChans, 5, 5, 5, 5, 5).to(device))

**Comment by pvjosue** [comment]
Actually in Pytorch the convention is that the first dimension is the batch, second the channels then the spatial dimensions. [batch,channel,dim1,dim2,...,dimN]
So when you have this tensor `1x8x3x3x3x3x3x3` you actually have batch 1, and 8 channels.
Instead of `trian_image_unsque = trian_image.unsqueeze(0)` try `trian_image_unsque = trian_image.unsqueeze(1)` to insert a single channel there.

**Comment by hnczwj2008** [comment]
wow, thank you very much  !
